tests/debugcommunity/node.py

```diff
@@ -266,214 +296,276 @@ def receive_message(self, addresses=None, names=None, timeout=0.5):
         assert names is None or isinstance(names, list), type(names)
         assert names is None or all(isinstance(name, unicode) for name in names), [type(name) for name in names]
 
-        for candidate, packet in self.receive_packet(addresses, timeout):
-            try:
-                message = self.decode_message(candidate, packet)
-            except ConversionNotFoundException as exception:
-                self._logger.exception("Ignored %s", exception)
-                continue
+        packets = yield self.receive_packet(addresses, timeout)
+        return_list = []
+        if packets:
+            for candidate, packet in packets:
+                try:
+                    message = yield self.decode_message(candidate, packet)
+                except ConversionNotFoundException as exception:
+                    self._logger.exception("Ignored %s", exception)
+                    continue
 
-            if not (names is None or message.name in names):
-                self._logger.debug("Ignored %s (%d bytes) from %s", message.name, len(packet), candidate)
-                continue
+                if not (names is None or message.name in names):
+                    self._logger.debug("Ignored %s (%d bytes) from %s", message.name, len(packet), candidate)
+                    continue
 
-            self._logger.debug("%s (%d bytes) from %s", message.name, len(packet), candidate)
-            yield candidate, message
+                self._logger.debug("%s (%d bytes) from %s", message.name, len(packet), candidate)
+                return_list.append((candidate, message))
+        returnValue(iter(return_list))
 
     @blocking_call_on_reactor_thread
     @inlineCallbacks
     def receive_messages(self, addresses=None, names=None, return_after=sys.maxint, timeout=0.5):
         messages = []
         for _ in xrange(5):
-            for message_tuple in self.receive_message(addresses, names, timeout):
-                messages.append(message_tuple)
-                if len(messages) == return_after:
+            received_messages = yield self.receive_message(addresses, names, timeout)
+            if received_messages:
+                for message_tuple in received_messages:
+                    messages.append(message_tuple)
+                    if len(messages) == return_after:
+                        break
+                if messages:
                     break
-            if messages:
-                break
             else:
                 # Wait for a bit and try again
-                yield deferLater(reactor, 0.005, lambda : None)
+                yield self.register_task("receive_messages_wait", deferLater(reactor, 0.005, lambda: None))
```

Comment:
I think this else statement needs an extra indent since it's part of the `if messages` check. Also, I think you can re move the `if received_messages` then.